### PR TITLE
Adds CMake support to wolfSSL

### PR DESCRIPTION
**Pull request by dgarske**
* Build wolfSSL as a library and builds all examples / tests.
* Added instructions in the INSTALL file.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+target_link_libraries(wolfcryptbench wolfssl)
+
+# Build wolfSSL Client example
+add_executable(client
+    ${CMAKE_CURRENT_SOURCE_DIR}/examples/client/client.c)
+target_link_libraries(client wolfssl)
+
+# Build wolfSSL Server example
+add_executable(server
+    ${CMAKE_CURRENT_SOURCE_DIR}/examples/server/server.c)
```
Comment (ArtBlnd) [inline CMakeLists.txt]: In mobile projects, such as android, iOS. this can cause build failure because this will be compiled as native apps(for example, on iOS. it will emit error that those executable need bundle identifier.)

Comment (ArtBlnd) [reply CMakeLists.txt]: What about make a flag not to compile bundle executables.

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+# Project
+####################################################
+project(wolfssl)
+find_package (Threads)
+
+####################################################
+# Compiler
+####################################################
+set(CMAKE_C_COMPILER gcc)
+set(CMAKE_CXX_COMPILER gcc)
```
Comment (ArtBlnd) [inline CMakeLists.txt]: This can cause build problem when making C++ mobile project that has package manager. something like conan.

**Comment by kaleb-himes** [comment]
PR issues resolved with new file test, I removed myself as a reviewer. @ArtBlnd has provided some helpful feedback.

File: CMakeLists.txt (hunk 2 above)
Comment (grooverdan) [reply CMakeLists.txt:21]: Agree. Omit these and let cmake use the default.

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+endif()
+
+####################################################
+# Build flags
+####################################################
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/.)
+include_directories(${CMAKE_CURRENT_BINARY_DIR}/.)
+
+# Example for map file and custom linker script
+set(CMAKE_C_FLAGS "-std=gnu89 ${CMAKE_C_FLAGS}")
```
Comment (grooverdan) [inline CMakeLists.txt:60]: is `-std=gnu89` required? For all files? if so should check compiler support first.

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    ${CMAKE_CURRENT_SOURCE_DIR}/examples/server/server.c
+    ${CMAKE_CURRENT_SOURCE_DIR}/examples/client/client.c)
+
+####################################################
+# Output Files
+####################################################
+# Build library
+add_library(wolfssl ${LIB_SOURCE_FILES})
+
+# Build wolfCrypt test
```
Comment (grooverdan) [inline CMakeLists.txt:81]: tSest components should be under a `IF(BUILD_TESTS='YES')` or similar

and an option to enable it like:

`SET(BUILD_TESTS "YES" CACHE  BOOL "Build with test")`

File: Makefile.am [hunk 5]
```diff
@@ -144,7 +144,9 @@ EXTRA_DIST+= INSTALL
 EXTRA_DIST+= IPP
 EXTRA_DIST+= LPCExpresso.cproject
 EXTRA_DIST+= LPCExpresso.project
-EXTRA_DIST+= resource.h wolfssl.rc
+EXTRA_DIST+= resource.h
```
Comment (grooverdan) [inline Makefile.am:147]: this doesn't look relevant to the cmake changes. Why is it here? If its needed maybe a separate PR.

File: wolfcrypt/src/evp.c [hunk 6]
```diff
@@ -2086,7 +2086,6 @@ int wolfSSL_EVP_read_pw_string(char* buf, int bufSz, const char* banner, int v)
     return 0;
 }
 #endif /* WOLFSSL_APACHE_HTTPD */
-#endif /* WOLFSSL_EVP_INCLUDED */
```
Comment (grooverdan) [inline wolfcrypt/src/evp.c:2089]: should be a separate PR if needed.

**Comment by dgarske** [comment]
Thanks @grooverdan and @ArtBlnd for your feedback. I'll work on integrating these suggestions shortly.

File: CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+#set(CMAKE_C_COMPILER "/opt/arm-linux-musleabihf-cross/bin/arm-linux-musleabihf-gcc")
+#set(CMAKE_CXX_COMPILER "/opt/arm-linux-musleabihf-cross/bin/arm-linux-musleabihf-g++")
+#set(CMAKE_SYSROOT "/opt/arm-linux-musleabihf-cross/arm-linux-musleabihf/")
+
+####################################################
+# Build Options
+####################################################
+if(EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/wolfssl/options.h")
+    # Copy generated ./options.h
+    file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/wolfssl/options.h
```
Comment (vaintroub) [inline CMakeLists.txt:35]: This is CONFIGURE_FILE, in cmake. it also retains the timestamp

File: CMakeLists.txt (hunk 3 above)
Comment (vaintroub) [reply CMakeLists.txt:60]: MariaDB builds without it, on at least 4 different compilers, some of which do not have this option

File: CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+file(GLOB TEST_SOURCE_FILES
+    ${CMAKE_CURRENT_SOURCE_DIR}/tests/*.c
+    ${CMAKE_CURRENT_SOURCE_DIR}/examples/server/server.c
+    ${CMAKE_CURRENT_SOURCE_DIR}/examples/client/client.c)
+
+####################################################
+# Output Files
+####################################################
+# Build library
+add_library(wolfssl ${LIB_SOURCE_FILES})
```
Comment (dgarske) [inline CMakeLists.txt:79]: Note: Also need to add the math lib... for DH builds.

```
# add the link_libraries(m) for linux
link_libraries(m)# Build library
```

**Comment by dgarske** [comment]
Improvements to CMake support based on feedback from users. Added "BUILD_TESTS" option and swiched to "configure_file".

**Comment by dgarske** [comment]
@toddouska : This has been reviewed and used by several customers. I've implemented all the feedback and this PR is in a good place for you to review.

File: CMakeLists.txt (hunk 8 above)
Comment (vaintroub) [reply CMakeLists.txt]: Usually, FIND_LIBRARY or CHECK_LIBRARY_EXISTS is used, also target_link_libraries() rather than link_libraries().
link_libraries(m) would not work on Windows. Also sometimes, on Unixes , libc contains math functionality.

File: CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+#
+# Copyright (C) 2006-2020 wolfSSL Inc.
+#
+# This file is part of wolfSSL. (formerly known as CyaSSL)
+#
+# Usage:
+# $ mkdir build
+# $ cd build
+# $ cmake ..
+# $ make
```
Comment (grooverdan) [inline CMakeLists.txt:11]: instead of `make`, `cmake -build .`. This works for all generators.

File: INSTALL [hunk 10]
```diff
@@ -73,3 +73,25 @@
...
+    Internally cmake is setup to do the following:
+    1. Uses the ./configure generated wolfssl/options.h as the build options by coping it to the build directory as user_settings.h.
+    2. Builds wolfSSL as library.
+    3. Builds the examples.
+
+    Build Steps:
+    $ mkdir build
+    $ cd build
+    $ cmake ..
+    $ make
```
Comment (grooverdan) [inline INSTALL:88]: `cmake --build .`

File: INSTALL [hunk 11]
```diff
@@ -73,3 +73,25 @@
...
+    $ sudo make install
```
Comment (grooverdan) [inline INSTALL:89]: [cmake  --install .](https://cmake.org/cmake/help/latest/manual/cmake.1.html#install-a-project)
